File: Leet-017. Letter Combinations of a Phone Number/solution-2^n.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.length() == 0) {
            return new LinkedList<>();
        }
        
        LinkedList<String> res = new LinkedList<>();
        res.add("");
        String[] dict = new String[] {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        
        for (int i = 0; i < digits.length(); i++) {
            String letters = dict[digits.charAt(i) - '0'];
            while (!res.isEmpty() && res.peek().length() == i) {
                String prev = res.remove();
                for (int j = 0; j < letters.length(); j++) {
                    res.add(prev + letters.charAt(j));
                }
            }
        }
        
        return res;
    }
}

File: Leet-017. Letter Combinations of a Phone Number/solution-2^n.js
/**
 * @param {string} digits
 * @return {string[]}
 */

// Time: O(2^n)
var letterCombinations = function(digits) {
    if (digits === null || digits.length === 0) return [];
    for (var i = 0; i < digits.length; i++) {
        if (digits[i] < '2' || digits[i] > '9') return [];
    }
    
    var map = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz'
    }
    
    var ret = [];
    combinations(digits, map, ret, []);
    
    return ret;
}

var combinations = function(digits, map, results, result) {
    if (digits.length === 0) {
        results.push(result.join(''));
        return ;
    }
    
    var letters = map[digits[0]];
    for (var i = 0; i < letters.length; i++) {
        result.push(letters[i]);
        combinations(digits.substring(1), map, results, result);
        result.pop();
    }
};